lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+      continue;
+    }
+    // Handle both phis and terminator results.
+    auto *bbArg = cast<SILPhiArgument>(incomingValue);
+    // The storage block must strictly dominate the phi.
+    if (!pass.domInfo->properlyDominates(
+          allocInst->getParent(), bbArg->getParent())) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void OpaqueStorageAllocation::allocatePhi(PhiValue phi) {
+  // Coalesces phi operand storage with the phi storage. The algorithm processes
+  // all incoming values at once, so it is is run when visiting the block
+  // argument.
+  //
+  // The phi operand projections are computed first to give them priority. Then
+  // we determine if the phi itself can share storage with one of its users.
+  CoalescedPhi coalescedPhi;
+  coalescedPhi.coalesce(phi, pass.valueStorageMap);
+
+  SmallVector<SILValue, 4> coalescedValues;
+  coalescedValues.resize(coalescedPhi.getCoalescedOperands().size());
+  for (SILValue value : coalescedPhi.getCoalescedValues())
+    coalescedValues.push_back(value);
+
+  if (!findPhiProjectionIntoUse(phi, coalescedValues))
+    createStackAllocationStorage(phi);
+
+  // Regardless of whether we projected into a user or allocated storage,
+  // provide this storage to all the incoming values that can reuse it.
+  for (Operand *phiOper : coalescedPhi.getCoalescedOperands()) {
+    removeAllocation(phiOper->get());
+    pass.valueStorageMap.recordPhiUseProjection(phiOper,
+                                                PhiOperand(phiOper).getValue());
+  }
+}
+
+// Unfortunately, we create alloc_stack instructions for SSA values before
+// coalescing block arguments. This temporary storage now needs to be removed.
+void OpaqueStorageAllocation::removeAllocation(SILValue value) {
+  auto &storage = pass.valueStorageMap.getStorage(value);
+  auto *allocInst = cast<AllocStackInst>(storage.storageAddress);
+  storage.storageAddress = nullptr;
+
+  // It's only use should be dealloc_stacks.
+  for (Operand *use : allocInst->getUses()) {
+    pass.deleter.forceDelete(cast<DeallocStackInst>(use->getUser()));
+  }
+  pass.deleter.forceDelete(allocInst);
+}
+
+// Create alloc_stack that dominates an owned value \p value. Create
+// jointly-postdominating dealloc_stack instructions.  Nesting will be fixed
+// later.
+//
+// Any value that may be used by a return instruction must be deallocated
+// immediately before the return. This allows the return to be rewritten by
+// loading from storage.
+AllocStackInst *OpaqueStorageAllocation::
+createStackAllocation(SILValue value) {
+  assert(value.getOwnershipKind() != OwnershipKind::Guaranteed && 
+         "creating storage for a guaranteed value implies a copy");
+
+#ifndef NDEBUG
+  // Instructions that produce an opened type never reach here because they
+  // have guaranteed ownership--they project their storage. We reach this
+  // point after the opened value has been copied.
+  if (auto *defInst = value->getDefiningInstruction()) {
```

Comment:
This if-statement is not needed, you can directly do
```
if (auto *singleValue = dyn_cast<SingleValueInstruction>(value))
```
or, even better:
```
assert((!isa<SingleValueInstruction>(value) || !cast<SingleValueInstruction>(value)->getDefinedOpenedArchetype()) && "...")
```
Then you don't need the `#ifndef NDEBUG`